virttest/nbd.py

```diff
@@ -0,0 +1,157 @@
+    elif unix_socket:
+        return 'nbd+unix://{export}?socket={sock}'.format(export=export_name,
+                                                          sock=unix_socket)
+
+    raise ValueError('Either nbd_server or nbd_unix_socket is required.')
+
+
+@error_context.context_aware
+def export_image(qemu_nbd, filename, local_image, params):
+    """
+    Export a local file image with qemu-nbd command
+    :param qemu_nbd: path to the qemu-nbd binary
+    :param filename: image path for raw/qcow2 image or image json
+                     representation string for luks image
+    :param local_image: image tag defined in parameter images
+    :param params: local image specified params
+    :return: pid of qemu-nbd server or None
+    """
+    cmd_dict = {
+        "secret_object": "",
+        "tls_creds": "",
+        "export_format": "",
+        "persistent": "-t",
+        "desc": "",
+        "port": "",
+        "export_name": "",
+        "unix_socket": "",
+        "filename": "",
+        "fork": "--fork",
+        "pid_file": ""
+    }
+    export_cmd = ('{secret_object} {tls_creds} '
+                  '{export_format} {persistent} {desc} {port} '
+                  '{export_name} {fork} {pid_file} {unix_socket} {filename}')
+
+    pid_file = utils_misc.generate_tmp_file_name('%s_nbd_server' % local_image,
+                                                 'pid')
+    cmd_dict['pid_file'] = '--pid-file %s' % pid_file
+    cmd_dict['filename'] = filename
+
+    # auto-detect format if format(-f) is not specified
+    if params.get('nbd_export_format'):
+        cmd_dict['export_format'] = '-f %s' % params['nbd_export_format']
+
+        if params['nbd_export_format'] == 'luks':
+            # We can only export a local luks image in luks
+            if params.get("image_format") != 'luks':
+                raise ValueError("Only a luks image can be exported in luks")
+
+            secret_str = '--object secret,id={aid},data={data}'
+            cmd_dict.update({
+                'secret_object': secret_str.format(
+                    aid='%s_encrypt0' % local_image,
+                    data=params["image_secret"]),
+                'filename': "'%s'" % filename
+            })
+
+    if params.get('nbd_export_name'):
+        cmd_dict['export_name'] = '-x %s' % params['nbd_export_name']
+
+    if params.get('nbd_export_description'):
+        cmd_dict['desc'] = '-D "%s"' % params['nbd_export_description']
+
+    if params.get('nbd_unix_socket'):
+        cmd_dict['unix_socket'] = '-k %s' % params['nbd_unix_socket']
+    else:
+        # 10809 is used by defalut if port is not set
+        if params.get('nbd_port'):
+            cmd_dict['port'] = '-p %s' % params['nbd_port']
+
+        # tls creds is supported for ip only
+        if params.get('nbd_server_tls_creds'):
+            tls_str = '--object tls-creds-x509,id={aid},endpoint=server,dir={tls_creds} --tls-creds {aid}'
```

Comment:
It's too long, please split them into multi-lines.

Reply:
change done